### PR TITLE
common/pick_address: Allow binding on loopback iface

### PR DESCRIPTION
**Pull request by tchaikov**
Fixes: https://tracker.ceph.com/issues/48893

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/common/pick_address.cc [hunk 1]
```diff
@@ -37,6 +38,75 @@
...
+    }
+    break;
+  }
+  return false;
+}
+
+bool matches_with_net(CephContext *cct,
+		      const ifaddrs& ifa,
+		      const std::string& s,
+                      unsigned ipv)
```
Comment (tchaikov) [inline src/common/pick_address.cc:82]: please be consistent when using tab and spaces for indent.

File: src/common/pick_address.cc [hunk 2]
```diff
@@ -37,6 +38,75 @@
...
+  if (boost::starts_with(ifa.ifa_name, "lo:")) {
+    return false;
+  }
+  return true;
+}
+
+bool matches_with_net(const ifaddrs& ifa,
+		      const sockaddr* net,
+		      unsigned int prefix_len,
+                      unsigned ipv)
```
Comment (tchaikov) [inline src/common/pick_address.cc:62]: tab and spaces

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-03-27_12:17:59-rados-wip-kefu-testing-2021-03-27-1545-distro-basic-gibba/
- https://pulpito.ceph.com/kchai-2021-03-28_03:01:25-rados-wip-kefu-testing-2021-03-27-1545-distro-basic-smithi/

failures

- https://tracker.ceph.com/issues/49344
- https://tracker.ceph.com/issues/49260
- https://tracker.ceph.com/issues/49633

**Comment by tchaikov** [comment]
@jdurgin @liewegas ping?

File: src/common/pick_address.cc [hunk 3]
```diff
@@ -45,9 +45,9 @@ bool matches_with_name(const ifaddrs& ifa, const std::string& if_name)
   return if_name.compare(ifa.ifa_name) == 0;
 }
 
-bool is_addr_allowed(const ifaddrs& ifa)
+bool is_addr_allowed(const ifaddrs& ifa, bool exclude_lo_iface)
 {
-  if (strcmp(ifa.ifa_name, "lo") == 0) {
+  if (exclude_lo_iface && strcmp(ifa.ifa_name, "lo") == 0) {
```
Comment (liewegas) [inline src/common/pick_address.cc:50]: what about ``lo:0`` etc?  should this check for ``lo`` or prefix ``lo:``?

**Review by liewegas — APPROVED** [review]
love the cleanup. i don't fully understand the reasons why we would or wouldn't want to use an IP on lo or lo:NNN, but putting it behind an option seems reasonable if it's dependent on a non-standard networking configuration.

File: src/common/pick_address.cc (hunk 3 above)
Comment (tchaikov) [reply src/common/pick_address.cc:50]: it is checked at line 53.

**Comment by tchaikov** [comment]
rebased against master.

File: src/common/pick_address.cc (hunk 3 above)
Comment (tchaikov) [reply src/common/pick_address.cc:50]: i think i am trying to address the issue fixed by #38925. in which, `lo` is allowed to be bound by daemons, while the `lo:` interfaces are not. and in the latest master, both `lo` and `lo:` are not allowed, so this option is introduced for the very use case of https://tracker.ceph.com/issues/48893.

**Pull request merged**